### PR TITLE
Attempt to reconnect if adapter disconnected

### PR DESCRIPTION
**Pull request by saimonmoore**
## What?

When a migration is doing something complex and takes a long time while not accessing the database, the current connection in the adapter may be disconnected causing later migrations to fail.

This PR attempts to reconnect 3 times in the case of a disconnection issue.

### CODE REVIEW

File: lib/lhm/connection.rb [hunk 1]
```diff
@@ -89,7 +89,17 @@ class ActiveRecordConnection
...
+      end
+
+      def set_database_name
+        retries        = 3
         @database_name = @adapter.current_database
+      rescue
+        retries -= 1
+        @adapter.reconnect!
+        puts "Reconnecting and retrying..." unless retries.zero?
+        retry unless (retries -= 1).zero?
```
Comment (somebody32) [inline lib/lhm/connection.rb]: to many `retries -= 1`?

File: lib/lhm/connection.rb [hunk 2]
```diff
@@ -89,7 +89,17 @@ class ActiveRecordConnection
...
+        set_database_name
+      end
+
+      def set_database_name
+        retries        = 3
         @database_name = @adapter.current_database
+      rescue
+        retries -= 1
+        @adapter.reconnect!
+        puts "Reconnecting and retrying..." unless retries.zero?
```
Comment (somebody32) [inline lib/lhm/connection.rb]: better to use one if for 2 statements

**Comment by saimonmoore** [comment]
thanks @somebody32 . Was very late last night...

**Comment by ivanilves** [comment]
:+1:

File: lib/lhm/connection.rb [hunk 3]
```diff
@@ -89,7 +89,18 @@ class ActiveRecordConnection
...
+
+      def set_database_name
+        retries        = 3
         @database_name = @adapter.current_database
+      rescue
+        unless (retries -= 1).zero?
+          @adapter.reconnect!
+          puts "Reconnecting and retrying..."
+          retry
+        end
```
Comment (somebody32) [inline lib/lhm/connection.rb]: do we need to reraise here? otherwise will be swallowed

Comment (saimonmoore) [inline lib/lhm/connection.rb]: hmm true...yep will reraise

**Comment by sauloperez** [comment]
Looks good :+1:

**Comment by enricostano** [comment]
reraise && :+1:

File: lib/lhm/connection.rb [hunk 4]
```diff
@@ -89,7 +89,21 @@ class ActiveRecordConnection
...
+        retries        = 3
+        begin
+          @database_name = @adapter.current_database
+        rescue => err
+          unless (retries -= 1).zero?
+            @adapter.reconnect!
+            puts "Reconnecting and retrying..."
+            retry
+          end
+          raise err
```
Comment (somebody32) [inline lib/lhm/connection.rb]: no need to pass an arg, will raise the last error by default

**Comment by somebody32** [comment]
will be great to squash commits and :+1: 
fell free to ignore my last comment if you want :)

File: lib/lhm/connection.rb [hunk 5]
```diff
@@ -89,7 +89,21 @@ class ActiveRecordConnection
...
-        @database_name = @adapter.current_database
+        set_database_name
+      end
+
+      def set_database_name
+        retries        = 3
+        begin
+          @database_name = @adapter.current_database
+        rescue => err
+          unless (retries -= 1).zero?
```
Comment (somebody32) [inline lib/lhm/connection.rb]: and then we can do an `early` raise here?
```ruby
rescue
  retries -= 1
  raise if retries.zero?

  @adapter.reconnect!
  puts "Reconnecting and retrying..."
  retry
```

**Comment by saimonmoore** [comment]
will squash...yep thanks guys

**Pull request merged**